phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixStatement.java

```diff
@@ -317,16 +332,112 @@ protected QueryPlan optimizeQuery(CompilableStatement stmt) throws SQLException
+        innerBuilder.setTenantId(ByteStringer.wrap(tenantId));
+        innerBuilder.setSchemaName(ByteStringer.wrap(schemaBytes));
+        innerBuilder.setTableName(ByteStringer.wrap(pTable.getTableName().getBytes()));
+        innerBuilder.setLastDDLTimestamp(pTable.getLastDDLTimestamp());
+    }
+    /**
+     * Build a request for the validateLastDDLTimestamp RPC.
+     * @param tableRef
+     * @return ValidateLastDDLTimestampRequest for the table in tableRef
+     */
+    private RegionServerEndpointProtos.ValidateLastDDLTimestampRequest getValidateDDLTimestampRequest(TableRef tableRef) throws TableNotFoundException {
+        RegionServerEndpointProtos.ValidateLastDDLTimestampRequest.Builder requestBuilder
+                = RegionServerEndpointProtos.ValidateLastDDLTimestampRequest.newBuilder();
+        RegionServerEndpointProtos.LastDDLTimestampRequest.Builder innerBuilder
+                = RegionServerEndpointProtos.LastDDLTimestampRequest.newBuilder();
+
+        byte[] tenantIDBytes = this.connection.getTenantId() == null
+                                    ? HConstants.EMPTY_BYTE_ARRAY
+                                    : this.connection.getTenantId().getBytes();
+        setLastDDLTimestampRequestParameters(innerBuilder, tenantIDBytes, tableRef.getTable());
+        requestBuilder.addLastDDLTimestampRequests(innerBuilder);
+
+        /*
+        when querying a view, we need to validate last ddl timestamps for all its ancestors.
+         */
+        if (tableRef.getTable().getType().equals(PTableType.VIEW)) {
+            PTable pTable = tableRef.getTable();
+            while (pTable.getParentName() != null) {
+                PTable parentTable = this.connection.getTable(new PTableKey(this.connection.getTenantId(), pTable.getParentName().getString()));
+                innerBuilder = RegionServerEndpointProtos.LastDDLTimestampRequest.newBuilder();
+                setLastDDLTimestampRequestParameters(innerBuilder, tenantIDBytes, parentTable);
+                requestBuilder.addLastDDLTimestampRequests(innerBuilder);
+                pTable = parentTable;
+            }
+        }
+        return requestBuilder.build();
+    }
+
+    /**
+     * Verifies that table metadata in client cache is up-to-date with server.
+     * A random live region server is picked for invoking the RPC to validate LastDDLTimestamp.
+     * Retry once if there was an error performing the RPC, otherwise throw the Exception.
+     * @param tableRef
+     * @throws SQLException
+     */
+    private void validateLastDDLTimestamp(TableRef tableRef, boolean doRetry) throws SQLException {
+
+        String infoString = getInfoString(tableRef);
+        try (Admin admin = this.connection.getQueryServices().getAdmin()) {
+            // if this is a retry, update the list of live region servers in CQSI
+            if (!doRetry) {
+                this.connection.getQueryServices().refreshLiveRegionServers();
```

Comment:
This seems little unusual to me. Normally whenever you encounter an exception, you need to clear the cache. So we should move refreshing live regionserver list call to the catch exception block.